### ISSUE TITLE
incorrect diagnostic when derive and where clauses on tuple structs are involved

### ISSUE THREAD

**Issue by gnzlbg** [rust-lang/rust#42259]
This code:

```rust
use generic_array::{ArrayLength, GenericArray};
use std::hash::{Hash};
use typenum::U2;

#[derive(Copy,Clone,Default,PartialEq,PartialOrd,Hash)]
pub struct Point<Num,AD>(GenericArray<Num, AD>)
    where
    Num: Copy + Default + PartialEq + PartialOrd + Hash,
    AD: ArrayLength<Num> + Copy + Default + PartialEq + PartialOrd + Hash,
    AD::ArrayType: Copy,
    GenericArray<Num, AD>: Copy
;

type Point2D = Point<f64, U2>;

#[derive(Copy,Clone,Default,PartialEq,PartialOrd,Hash)]
struct Vertex2D {
    position: Point2D,
    value: f32,
}
```

produces this error:

```shell
error[E0204]: the trait `Copy` may not be implemented for this type
  --> src/geometry.rs:17:10
   |
17 | #[derive(Copy,Clone,Default,PartialEq,PartialOrd,Hash)]
   |          ^^^^
18 | struct Vertex2D {
19 |     position: Point2D,
   |     ----------------- this field does not implement `Copy`

error: aborting due to previous error
```

The error is, however, that `f64` does not implement `Hash`. This code compiles fine: 

```shell
use generic_array::{ArrayLength, GenericArray};
use typenum::U2;

#[derive(Copy,Clone,Default,PartialEq,PartialOrd)]
pub struct Point<Num,AD>(GenericArray<Num, AD>)
    where
    Num: Copy + Default + PartialEq + PartialOrd,
    AD: ArrayLength<Num> + Copy + Default + PartialEq + PartialOrd,
    AD::ArrayType: Copy,
    GenericArray<Num, AD>: Copy
;

type Point2D = Point<f64, U2>;

#[derive(Copy,Clone,Default,PartialEq,PartialOrd)]
struct Vertex2D {
    position: Point2D,
    value: f32,
}
```

The proper error message here should be that `Point<Num, AD>` requires `Num: Hash` and `f64` does not implement `Hash`, pointing out to the line `type Point2D = Point<f64, U2>`. 

Instead, the error points to the wrong line (`position` inside `Vertex`), and also the reason is wrong. It says that `position` does not implement `Copy`,  but this is not possible because `Point` derives `Copy` and its fields are constrained to derive `Copy` as well, so deriving `Copy` for `Point` cannot fail. This error message is extremely baffling. 

@rkruppe provides the following mwe:

```rust
// try uncommenting the where clause:
#[derive(Copy, Clone)]
struct Foo<T>(T) /*where T: IntoIterator<Item=u8>*/;

fn assert_copy<T: Copy>() {}

fn main() {
    assert_copy::<Foo<i32>>();
}
```

**Comment by rkruppe** [comment]
Worth pointing out that my reduction at least mentions the `i32: IntoIterator` obligation, while in the bigger example there's no hint at all.

**Comment by gnzlbg** [comment]
It seems that there are two independent failures here:

1.  No diagnostic for `type Point2D = Point<f64, U2>;` when `Point<Num: Hash, ...>`. Since `f64: !Hash`, this can't ever succeed.

2. Insufficient diagnostic for errors in derived impls. The where clauses of the tuple struct apply to all derived impls. That is, the derived `Copy` impl applies only for types that satisfy the whole where clause of the tuple struct. This is not the case for `f64` since it does not implement `Hash`. Hence, the resulting `Point` type is not `Copy`. Ideally, the diagnostic should say that there is a generic impl of `Copy` for `Point` (which is derived in this case) that fails  because `f64` is not `Hash`.

**Comment by KyleMayes** [comment]
This is still an issue as of the latest nightly, just spent more time than I would like to admit trying to figure out why a type wasn't `Copy`. The following self-contained example showcases the error:

[Rust Playground](https://play.rust-lang.org/?gist=2792e19325c4db84207d10c647af827e&version=undefined)

```rust
use std::marker::{PhantomData};

pub trait Trait { }

impl Trait for i32 { }

#[derive(Copy, Clone)]
pub struct Struct<T> where T: Trait {
    _marker: PhantomData<*const T>,
}

#[derive(Copy, Clone)]
pub struct DoesntWork {
    a: Struct<i32>,
    b: Struct<i64>,
}

fn main() { }
```

Removing the `#[derive(Copy, Clone)]` above the `DoesntWork` struct results in the expected error message, (`the trait bound ``i64: Trait`` is not satisfied`), but instead you get: 

```
error[E0204]: the trait `Copy` may not be implemented for this type
  --> src/main.rs:12:10
   |
12 | #[derive(Copy, Clone)]
   |          ^^^^
...
15 |     b: Struct<i64>,
   |     -------------- this field does not implement `Copy`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0204`.
error: Could not compile `playground`.

To learn more, run the command again with --verbose.
```

**Comment by estebank** [comment]
Triage: no change.

**Comment by estebank** [comment]
Current output:

```
error[E0204]: the trait `Copy` may not be implemented for this type
  --> src/main.rs:12:10
   |
12 | #[derive(Copy, Clone)]
   |          ^^^^
...
15 |     b: Struct<i64>,
   |     -------------- this field does not implement `Copy`
   |
note: the `Copy` impl for `Struct<i64>` requires that `i64: Trait`
  --> src/main.rs:15:8
   |
15 |     b: Struct<i64>,
   |        ^^^^^^^^^^^
   = note: this error originates in the derive macro `Copy` (in Nightly builds, run with -Z macro-backtrace for more info)
```
